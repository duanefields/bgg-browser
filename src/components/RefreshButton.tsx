import { faSync } from "@fortawesome/free-solid-svg-icons"
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
import { IconButton } from "@mui/material"
import { useIsFetching, useQueryClient } from "@tanstack/react-query"

type RefreshButtonProps = {
  username: string
}

const RefreshButton = ({ username }: RefreshButtonProps) => {
  const queryClient = useQueryClient()
  const isFetching = useIsFetching()

  const onClick = () => {
    void queryClient.invalidateQueries({ queryKey: ["collection", username] })
  }

  return (
    <IconButton aria-label="refresh" onClick={onClick}>
      <FontAwesomeIcon icon={faSync} spin={isFetching > 0} />
    </IconButton>
  )
}

export default RefreshButton
