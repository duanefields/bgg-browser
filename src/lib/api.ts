/** Root URL of the BGG Proxy Server */
const BGG_PROXY: string = import.meta.env.VITE_BGG_PROXY

type CollectionResponse = {
  items: {
    item: {
      _objectid: number
      name: {
        _sortindex: string
        __text: string
      }
      yearpublished: string
      image: string
      thumbnail: string
      stats: {
        _minplayers: string
        _maxplayers: string
        _minplaytime: string
        _maxplaytime: string
        _playingtime: string
        _numowned: string
        rating: {
          usersrated: {
            _value: string
          }
          average: {
            _value: string
          }
          bayesaverage: {
            _value: string
          }
          stddev: {
            _value: string
          }
          median: {
            _value: string
          }
          ranks: {
            rank: {
              _type: string
              _id: string
              _name: string
              _friendlyname: string
              _value: string
              _bayesaverage: string
            }[]
            _value: string
          }
        }
      }
      status: {
        _own: string
        _prevowned: string
        _fortrade: string
        _want: string
        _wanttoplay: string
        _wanttobuy: string
        _wishlist: string
        _preordered: string
        _lastmodified: string
      }
      numplays: string
      _objecttype: string
      _subtype: string
      _collid: string
    }[]
  }
  _totalitems: string
  _termsofuse: string
  _pubdate: string
}

type Game = {
  objectId: number
  name: string
  thumbnail: string
  minPlayers: number
  maxPlayers: number
  minPlaytime: number
  maxPlaytime: number
  playtime: number
  yearPublished: number
  averageRating: number
  // rank: number
  numPlays: number
}

const getCollection = async (username: string): Promise<Game[]> => {
  const response = await fetch(
    `${BGG_PROXY}/collection/?username=${username}&stats=1&own=1&version=1&excludesubtype=boardgameexpansion`,
  )

  // Check for errors, the server responds with a 200 OK even if there is an error
  const json = await response.json()
  if (json.errors) {
    throw new Error(json.errors.error.message)
  }

  const data = json as CollectionResponse
  return data.items.item.map((item) => ({
    objectId: Number(item._objectid),
    name: item.name.__text,
    thumbnail: item.thumbnail,
    minPlayers: Number(item.stats._minplayers || 0),
    maxPlayers: Number(item.stats._maxplayers || 0),
    minPlaytime: Number(item.stats._minplaytime || 0),
    maxPlaytime: Number(item.stats._maxplaytime || 0),
    playtime: Number(item.stats._playingtime || 0),
    yearPublished: Number(item.yearpublished),
    averageRating: Number(Number(item.stats.rating.average._value || 0).toFixed(1)),
    // rank: item.stats.rank,
    numPlays: Number(item.numplays),
  }))
}

export { getCollection, BGG_PROXY }
