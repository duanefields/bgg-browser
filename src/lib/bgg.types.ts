export interface RankResponse {
  _type: string
  _id: string
  _name: string
  _friendlyname: string
  _value: string
  _bayesaverage: string
}

export interface ItemResponse {
  _objectid: number
  name: {
    _sortindex: string
    __text: string
  }
  yearpublished: string
  image: string
  thumbnail: string
  stats: {
    _minplayers: string
    _maxplayers?: string
    _minplaytime?: string
    _maxplaytime?: string
    _playingtime?: string
    _numowned: string
    rating: {
      usersrated: {
        _value: string
      }
      average: {
        _value: string
      }
      bayesaverage: {
        _value: string
      }
      stddev: {
        _value: string
      }
      median: {
        _value: string
      }
      ranks: {
        rank: RankResponse[] | { _value: string }
        _value: string
      }
      _value: string
    }
  }
  status: {
    _own: string
    _prevowned: string
    _fortrade: string
    _want: string
    _wanttoplay: string
    _wanttobuy: string
    _wishlist: string
    _preordered: string
    _lastmodified: string
  }
  numplays: string
  _objecttype: string
  _subtype: string
  _collid: string
}

export interface CollectionResponse {
  items: {
    item?: ItemResponse[]
  }
  _totalitems: string
  _termsofuse: string
  _pubdate: string
}

export interface UserResponse {
  user: {
    firstname: { _value: string }
    lastname: { _value: string }
    avatarlink: { _value: string }
    yearregistered: { _value: string }
    lastlogin: { _value: string }
    stateorprovince: { _value: string }
    country: { _value: string }
    webaddress: { _value: string }
    xboxaccount: { _value: string }
    wiiaccount: { _value: string }
    psnaccount: { _value: string }
    battlenetaccount: { _value: string }
    steamaccount: { _value: string }
    traderating: { _value: string }
    _id: string
    _name: string
    _termsofuse: string
  }
}

export interface InvalidUsernameResponse {
  errors: {
    error: {
      message: string
    }
  }
}

export interface CollectionProcessingResponse {
  message: string
}
