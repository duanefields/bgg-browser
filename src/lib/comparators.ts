import { Game } from "../shared.types"

/** sort by name ascending, ignoring leading articles a, an, and the */
export const titleComparator = (a: Game, b: Game) => {
  const aTitle = a.name.replace(/^(a|an|the) /i, "")
  const bTitle = b.name.replace(/^(a|an|the) /i, "")
  return aTitle.localeCompare(bTitle)
}

/** sort by averageRating descending, then by title */
export const ratingComparator = (a: Game, b: Game) => {
  return b.averageRating - a.averageRating || titleComparator(a, b)
}

/** sort by myRating descending, then by title */
export const myRatingComparator = (a: Game, b: Game) => {
  return (b.myRating || 0) - (a.myRating || 0) || titleComparator(a, b)
}

/** sort by numPlays descending, then by title */
export const playsComparator = (a: Game, b: Game) => {
  return b.numPlays - a.numPlays || titleComparator(a, b)
}

/** sort by rank ascending, then by title (though rank should be unique) */
export const rankComparator = (a: Game, b: Game) => {
  return (
    (a.rank || Number.MAX_SAFE_INTEGER) - (b.rank || Number.MAX_SAFE_INTEGER) ||
    titleComparator(a, b)
  )
}
