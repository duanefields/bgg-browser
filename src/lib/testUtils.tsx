import { QueryClient, QueryClientProvider } from "@tanstack/react-query"
import { createRootRoute, createRoute, createRouter, RouterProvider } from "@tanstack/react-router"
import { render } from "@testing-library/react"

/** Renders a collection of children with a Route and QueryClient Provider */
export const renderWithProviders = (children: React.ReactNode) => {
  // wrap the component in a fresh QueryClientProvider to isolate the tests
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  })

  // create a router with a single route that renders the children in a QueryClientProvider
  const rootRoute = createRootRoute()
  const indexRoute = createRoute({
    getParentRoute: () => rootRoute,
    path: "/",
    component: () => <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>,
  })
  const router = createRouter({
    routeTree: rootRoute.addChildren([indexRoute]),
  })

  return render(<RouterProvider router={router} />)
}

/** Renders a collection of children with just a QueryClient Provider */
export const renderWithQueryProvider = (children: React.ReactNode) => {
  // wrap the component in a fresh QueryClientProvider to isolate the tests
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  })

  return render(<QueryClientProvider client={queryClient}>{children}</QueryClientProvider>)
}
