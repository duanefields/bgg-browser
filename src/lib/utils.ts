import { Game } from "../shared.types"

/** Randomly shuffle an array of games */
export const shuffle = (games: Game[]) => {
  return games
    .map((value) => ({ value, sort: Math.random() }))
    .sort((a, b) => a.sort - b.sort)
    .map(({ value }) => value)
}
