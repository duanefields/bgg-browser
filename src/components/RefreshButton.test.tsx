import * as reactQuery from "@tanstack/react-query"
import { screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import { afterEach, expect, it, vi } from "vitest"
import { renderWithQueryProvider } from "../lib/testUtils"

import RefreshButton from "./RefreshButton"

const mockInvalidateQueries = vi.fn()

afterEach(() => {
  vi.restoreAllMocks()
})

it("Should match snapshot", () => {
  vi.spyOn(reactQuery, "useIsFetching").mockReturnValue(1)
  const view = renderWithQueryProvider(<RefreshButton username="test" />)
  expect(view).toMatchSnapshot()
})

it("Should render a refresh button", () => {
  renderWithQueryProvider(<RefreshButton username="test" />)
  expect(screen.getByRole("button")).toBeVisible()
})

it("Should spin the icon when fetching", async () => {
  const mockUseIsFetching = vi.spyOn(reactQuery, "useIsFetching").mockReturnValue(1)
  renderWithQueryProvider(<RefreshButton username="test" />)
  expect(mockUseIsFetching).toHaveBeenCalledOnce()
  expect(await screen.findByRole("img", { hidden: true })).toHaveClass("fa-spin")
})

it("Should not spin the icon when not fetching", async () => {
  const mockUseIsFetching = vi.spyOn(reactQuery, "useIsFetching").mockReturnValue(0)
  renderWithQueryProvider(<RefreshButton username="test" />)
  expect(mockUseIsFetching).toHaveBeenCalledOnce()
  expect(await screen.findByRole("img", { hidden: true })).not.toHaveClass("fa-spin")
})

it("Should call the queryClient.invalidateQueries function when clicked", async () => {
  vi.mock("@tanstack/react-query", async () => {
    const reactQuery = await vi.importActual("@tanstack/react-query")
    return { ...reactQuery, useQueryClient: () => ({ invalidateQueries: mockInvalidateQueries }) }
  })
  renderWithQueryProvider(<RefreshButton username="test" />)
  const user = userEvent.setup()
  await user.click(screen.getByRole("button"))
  expect(mockInvalidateQueries).toHaveBeenCalledWith({ queryKey: ["collection", "test"] })
})
