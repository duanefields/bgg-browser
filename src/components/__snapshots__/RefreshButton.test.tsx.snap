// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        aria-label="refresh"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-arrows-rotate fa-spin "
          data-icon="arrows-rotate"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M105.1 202.6c7.7-21.8 20.2-42.3 37.8-59.8c62.5-62.5 163.8-62.5 226.3 0L386.3 160 352 160c-17.7 0-32 14.3-32 32s14.3 32 32 32l111.5 0c0 0 0 0 0 0l.4 0c17.7 0 32-14.3 32-32l0-112c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 35.2L414.4 97.6c-87.5-87.5-229.3-87.5-316.8 0C73.2 122 55.6 150.7 44.8 181.4c-5.9 16.7 2.9 34.9 19.5 40.8s34.9-2.9 40.8-19.5zM39 289.3c-5 1.5-9.8 4.2-13.7 8.2c-4 4-6.7 8.8-8.1 14c-.3 1.2-.6 2.5-.8 3.8c-.3 1.7-.4 3.4-.4 5.1L16 432c0 17.7 14.3 32 32 32s32-14.3 32-32l0-35.1 17.6 17.5c0 0 0 0 0 0c87.5 87.4 229.3 87.4 316.7 0c24.4-24.4 42.1-53.1 52.9-83.8c5.9-16.7-2.9-34.9-19.5-40.8s-34.9 2.9-40.8 19.5c-7.7 21.8-20.2 42.3-37.8 59.8c-62.5 62.5-163.8 62.5-226.3 0l-.1-.1L125.6 352l34.4 0c17.7 0 32-14.3 32-32s-14.3-32-32-32L48.4 288c-1.6 0-3.2 .1-4.8 .3s-3.1 .5-4.6 1z"
            fill="currentColor"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      aria-label="refresh"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-arrows-rotate fa-spin "
        data-icon="arrows-rotate"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M105.1 202.6c7.7-21.8 20.2-42.3 37.8-59.8c62.5-62.5 163.8-62.5 226.3 0L386.3 160 352 160c-17.7 0-32 14.3-32 32s14.3 32 32 32l111.5 0c0 0 0 0 0 0l.4 0c17.7 0 32-14.3 32-32l0-112c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 35.2L414.4 97.6c-87.5-87.5-229.3-87.5-316.8 0C73.2 122 55.6 150.7 44.8 181.4c-5.9 16.7 2.9 34.9 19.5 40.8s34.9-2.9 40.8-19.5zM39 289.3c-5 1.5-9.8 4.2-13.7 8.2c-4 4-6.7 8.8-8.1 14c-.3 1.2-.6 2.5-.8 3.8c-.3 1.7-.4 3.4-.4 5.1L16 432c0 17.7 14.3 32 32 32s32-14.3 32-32l0-35.1 17.6 17.5c0 0 0 0 0 0c87.5 87.4 229.3 87.4 316.7 0c24.4-24.4 42.1-53.1 52.9-83.8c5.9-16.7-2.9-34.9-19.5-40.8s-34.9 2.9-40.8 19.5c-7.7 21.8-20.2 42.3-37.8 59.8c-62.5 62.5-163.8 62.5-226.3 0l-.1-.1L125.6 352l34.4 0c17.7 0 32-14.3 32-32s-14.3-32-32-32L48.4 288c-1.6 0-3.2 .1-4.8 .3s-3.1 .5-4.6 1z"
          fill="currentColor"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
